webapp/sources/rudder/rudder-core/src/main/scala/com/normation/rudder/apidata/JsonQueryObjects.scala

```diff
@@ -0,0 +1,602 @@
+*/
+
+package com.normation.rudder.apidata
+
+import com.normation.GitVersion
+import com.normation.GitVersion.ParseRev
+import com.normation.GitVersion.Revision
+import com.normation.cfclerk.domain.TechniqueName
+import com.normation.cfclerk.domain.TechniqueVersion
+import com.normation.errors.PureResult
+import com.normation.errors.Unexpected
+import com.normation.rudder.domain.policies._
+import com.normation.rudder.rule.category.RuleCategory
+import com.normation.rudder.rule.category.RuleCategoryId
+import net.liftweb.common._
+import net.liftweb.http.Req
+import zio.json._
+import com.normation.errors._
+import com.normation.rudder.domain.properties.GlobalParameter
+import com.typesafe.config.ConfigValue
+import com.normation.rudder.domain.nodes.NodeGroup
+import com.normation.rudder.domain.nodes.NodeGroupCategoryId
+import com.normation.rudder.domain.nodes.NodeGroupId
+import com.normation.rudder.domain.properties.CompareProperties
+import com.normation.rudder.domain.properties.GenericProperty
+import com.normation.rudder.domain.properties.GroupProperty
+import com.normation.rudder.domain.properties.InheritMode
+import com.normation.rudder.domain.properties.PatchProperty
+import com.normation.rudder.domain.properties.PropertyProvider
+import com.normation.rudder.domain.queries.NodeReturnType
+import com.normation.rudder.domain.queries.QueryReturnType
+import com.normation.rudder.domain.queries.QueryTrait
+import com.normation.rudder.services.queries.CmdbQueryParser
+import com.normation.rudder.services.queries.JsonQueryLexer
+import com.normation.rudder.services.queries.StringCriterionLine
+import com.normation.rudder.services.queries.StringQuery
+import io.scalaland.chimney.dsl._
+
+/*
+ * This class deals with everything serialisation related for API.
+ * Change things with care! Everything must be versionned!
+ * Even changing a field name can lead to an API incompatible change and
+ * so will need a new API version number (and be sure that old behavior is kept
+ * for previous versions).
+ */
+
+
+///////////////////////////// decoder ///////////////
+
+/*
+ * Object used in JSON query POST/PUT request.
+ * For disambiguation, objects are prefixed by JQ.
+ *
+ * Note about id/revision: it does not make sense to try to change a
+ * revision, it's immutable. Only head can be changed, so rev is never
+ * specified in JQ object.
+ * It can make sense to clone a specific revision, so sourceId can have a
+ * sourceRevision too.
+ */
+object JsonQueryObjects {
+  import JsonResponseObjects.JRRuleTarget
+
+  final case class JQRuleCategory(
+        name       : Option[String] = None
+      , description: Option[String] = None
+      , parent     : Option[String] = None
+      , id         : Option[String] = None
+    ) {
+
+    def update(ruleCategory: RuleCategory) = {
+      ruleCategory.using(this).ignoreRedundantPatcherFields.patch
+    }
+  }
+
+
+  final case class JQDirectiveSectionVar(
+      name : String
+    , value: String
```

Comment:
we used to have null value and "" value, having sometimes different or similar behaviour.
shouldn't the value be an option for empty value that are not mandatory?

Reply:
I think we just don't have that level of precision from the user input. The parsing is done afterward, here the object is just a direct translation of json. It means that for optionnal values, the user can't do:
```
{"vars": [ {"var": { "name":"var1", "value":"val1" }, {"var": { "name":"var2" }}]}
```
And should do: 
```
{"vars": [ {"var": { "name":"var1", "value":"val1" }}]}
```
(and var2 will be missing on parsing, and it will be set to None in the domain object). 

If you know about a precise use case, I can add it in test in an other ticket.

Reply:
no, tht's fine like that: if it's the user input, it is as it should be. Thank you for the explaination